obs2ioda-v2/src/f_c_string_1D_t_mod.f90

```diff
@@ -0,0 +1,121 @@
+! Module providing functionality for handling one-dimensional arrays of
+! Fortran strings as C-compatible null-terminated strings.
+module f_c_string_1D_t_mod
+    use iso_c_binding, only : c_loc, c_ptr, c_null_char, c_char, c_f_pointer, c_null_ptr, &
+            c_associated
+    use f_c_string_t_mod, only : f_c_string_t, c_strlen
+    implicit none
+
+    ! Type to handle a 1D array of Fortran strings and convert them to/from
+    ! C-compatible null-terminated strings.
+    !
+    ! Array of f_c_string_t, each representing a null-terminated C string.
+    ! type(f_c_string_t), allocatable :: f_c_string_t_array(:)
+    !
+    ! Array of C pointers, each pointing to a null-terminated string in memory.
+    ! type(c_ptr), allocatable :: fc_string_1D(:)
+    type :: f_c_string_1D_t
+        ! Allocatable C-compatible null-terminated string
+        type(f_c_string_t), allocatable :: f_c_string_t_array(:)
+        type(c_ptr), allocatable :: fc_string_1D(:)
+
+    contains
+
+        procedure :: to_c => to_c
+        procedure :: to_f => to_f
+        final :: cleanup
+    end type f_c_string_1D_t
+
+contains
+
+    ! Converts a Fortran array of strings to a C-compatible null-terminated
+    ! string array and stores it internally as `fc_string_1D`.
+    !
+    ! Arguments:
+    ! - this: The instance of f_c_string_1D_t being operated on.
+    ! - f_string_1D: A 1D Fortran array of character strings.
+    !
+    ! Returns:
+    ! - c_string_1D: A C pointer to an array of C pointers, where
+    !       each pointer in the array points to a C string.
+    function to_c(this, f_string_1D) result(c_string_1D)
+        class(f_c_string_1D_t), target, intent(inout) :: this
+        character(len = *), intent(in) :: f_string_1D(:)
+        character(len = :), allocatable :: f_string
+        type(c_ptr) :: c_string_1D
+        integer :: i, m
+        m = size(f_string_1D)
+        if (allocated(this%fc_string_1D)) then
+            deallocate(this%fc_string_1D)
+        end if
+        if (allocated(this%f_c_string_t_array)) then
+            deallocate(this%f_c_string_t_array)
+        end if
+        allocate(this%f_c_string_t_array(m))
+        allocate(this%fc_string_1D(m))
+        do i = 1, m
+            f_string = f_string_1D(i)
+            this%fc_string_1D(i) = this%f_c_string_t_array(i)%to_c(f_string)
```

Comment:
I think it might be possible to avoid the use of an intermediate `f_string` variable here, and to call `to_c` directly on `f_string_1D(i)`.